tensornetwork/matrixproductstates/mps.py

```diff
@@ -0,0 +1,564 @@
+  to `len(FiniteMPS) - 1`. If the shift is a "right-shift" (i.e. 
+  `center_position` is moved from left to right), then all sites that are 
+  visited in between are left in left-orthogonal form. If the shift is a 
+  "right-shift" (i.e. `center_position` is shifted from right to left), 
+  then all sites that are visited in between are left in right-orthogonal form. 
+  For random initial tensors `tensors`, doing one sweep from left to right 
+  and a successive sweep from right to left brings the state into 
+  central canonical form. In this state, all sites to the left of 
+  `center_position` are left orthogonal, and all sites to the right of 
+  `center_position` are right orthogonal, and the state is normalized.
+  Due to efficiency reasons, the state upon initialization is usually 
+  NOT brought into the central canonical form.
+  """
+
+  def __init__(self,
+               tensors: List[Tensor],
+               center_position: int,
+               backend: Optional[Text] = None,
+               dtype: Optional[Type[np.number]] = None) -> None:
+    """
+    Initialize a FiniteMPS.
+    Args:
+      tensors: A list of `Tensor` objects.
+      center_position: The initial position of the center site.
+      backend: The name of the backend that should be used to perform 
+        contractions. See documentation of TensorNetwork.__init__ for 
+        a list of supported backends.
+      dtype: An optional `dtype` for the FiniteMPS. See documentation of 
+        TensorNetwork.__init__ for more details.
+    Returns:
+      None
+
+    """
+    if center_position < 0 or center_position >= len(tensors):
+      raise ValueError(
+          'center_position = {} not between 0 <= center_position < {}'.format(
+              center_position, len(tensors)))
+    super().__init__(backend=backend, dtype=dtype)
+    self.nodes = [
+        self.add_node(tensors[n], name='node{}'.format(n))
+        for n in range(len(tensors))
+    ]  #redundant?!
+    for site in range(len(self.nodes) - 1):
+      self.connect(self.nodes[site][2], self.nodes[site + 1][0])
+    self.center_position = center_position
+
+  def save(self, path: str):
+    raise NotImplementedError()
+
+  @property
+  def D(self) -> List:
+    """
+    Return a list of bond dimensions of FiniteMPS
+    """
+    return [self.nodes[0].shape[0]] + [node.shape[2] for node in self.nodes]
+
+  @property
+  def d(self) -> List:
+    """
+    Return a list of physical Hilbert-space dimensions of FiniteMPS
+    """
+
+    return [node.shape[1] for node in self.nodes]
+
+  def position(self, site: int, normalize: Optional[bool] = True) -> np.number:
+    """
+    Shift FiniteMPS.center_position to `site`.
+    Args:
+      site: The site to which FiniteMPS.center_position should be shifted
+      normalize: If `True`, normalize matrices when shifting.
+    Returns:
+      Tensor: The norm of the tensor at FiniteMPS.center_position
+    """
+
+    if site >= len(self.nodes) or site < 0:
```

Comment:
Could you add comments to this code? (And throughout the rest of this file.)

Reply:
Done!